### ISSUE TITLE
SeSAMe

### ISSUE THREAD

**Issue by zwdzwd** [Bioconductor/Contributions#716]
Hi, @vobencha ,
This is the first submission of the data package for sesame. Per our discussion, I closed the previous issue of this package and resubmit the data package here first. I will submit the code package and post my response to your last review once the data package is approved. Thank you for your help!

Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/zwdzwd/sesameData

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by zwdzwd** [comment]
Dear @bioc-issue-bot ,

I have previous posted the repository in #685. The old issue was closed and I opened a new issue to allow the addition of a depended data package (sesameData). Please link this repository to #716 . 

Thank you very much for the help!

**Comment by lshep** [comment]
**sesameData Package:** 
- [ ] Are all of these necessary and not already provided in a currently available Bioconductor annotation package?  or available in the AnnotationHub? 
- [ ]  Could you please provide a /inst/scripts directory with scripts (can be sudo code) describing how the data objects in /inst/extdata were created and any pertinent information regarding that data. 
- [ ] Please provide more information for the data man files regarding how the files were created, what is the original source of the files, etc.  

**sesame Package:**
Thank you for implementing the already suggested changes. 
Some other notes: 

NAMESPACE
- [ ] are all these exports really necessary or are some internal functions that would not be called by the user?  Only export functions that would directly be called by the user. 

R CODE 
- [ ] setMethod( "show", "SigSet",  - you don't need multiple cat functions - one cat can handle formatting through "\n" utilization
- [ ] have stronger parameter checking  on main functions and functions that have specific limited values. 
For example  makeExampleSeSAMeDataSet , it seems like the value of platform can only be HM450, EPIC or HM27   so having the arugment `platform = c("HM450", "EPIC", "HM27")`  and then use platform <- match.arg(platform)` in the code .   More general, provide some initial checking if the input to your main functions is the expected object type.  This should be applied throughout the entire code. (ie. stopifnot)  

Thank you for your submission.  I think the code is generally in really good shape and on track to be accepted before the April 11th deadline.  Please address the above issues in each of the packages and perform the appropriate version bump to have a rebuild of the package.  Please also comment back here with further clarification on what has been addressed. 

Cheers

**Comment by zwdzwd** [comment]
Thank you for the thorough review and encouraging comments! Your comments are addressed below, italicized. Changes are reflected in the latest updates. Let me know if I missed anything. 

**sesameData Package:**

- Are all of these necessary and not already provided in a currently available Bioconductor annotation package? or available in the AnnotationHub?

_Most of the curation data was first submitted to BioC and the original was provided in http://zwdzwd.github.io/InfiniumAnnotation. Although some part of the detail contained in this curation can overlap with existing data package such as `IlluminaHumanMethylation450kprobe`, our annotation adds critical correction to the official manifest and necessary additional information (such as the probe location on hg38, and cross-hybridization potential evaluation, binarized reference methylation status for a few cell types, pseudo-autosomal probes in sex chromosomes and extension base identity) for the proper functioning of the sesame package. Much content was detailed in our data curation paper (PMID 27924034). Part of the data included in sesameData also provide models we trained for certain sesame functions (such as inferEthnicity etc).  My colleague and I were in charge of curating these data. We made some attempt adopting existing packages but either found they do not suit our direct need. I will keep on studying relevant packages and try to utilize more existing standard annotations. But I hope current approach of providing an independent clean version gives the package better performance and make it perhaps less error-prone. Description of the data sources has been added to reflect the novelty of the included data (below)._

- Could you please provide a /inst/scripts directory with scripts (can be sudo code) describing how the data objects in /inst/extdata were created and any pertinent information regarding that data.

_Thanks for the suggestion. I created a file named `make-data.R` under `inst/scripts/` folder and added details describing how each piece of the data was obtained, the source (sample IDs, GEO IDs) and scripts showing how the data was preprocessed (background subtraction, genome segmentation etc.)._

- Please provide more information for the data man files regarding how the files were created, what is the original source of the files, etc.

_Thanks for the suggestion. I expanded the man pages for the data/*.rda and included the original data source (GEO experiment IDs or the URLs pointing to the raw data). For processed data, I included scripts that produce the data in make-data.R, for example the script to fit the ethnicity/sex inference model and the script to generate the reference methylation status vectors. For some (e.g, leukocyte reference methylation and the reference methylation of other cell types/tissue types), I also included brief description of the rationale behind the data processing in the man page._

**sesame Package:**
Thank you for implementing the already suggested changes.
Some other notes:
NAMESPACE
are all these exports really necessary or are some internal functions that would not be called by the user? Only export functions that would directly be called by the user.

_Thanks for the reminder. I double-checked all the exported functions to make sure they represent a direct interface of the package to the user._

R CODE
- setMethod( "show", "SigSet", - you don't need multiple cat functions - one cat can handle formatting through "\n" utilization

_I agree and I removed redundant cat functions._

- have stronger parameter checking on main functions and functions that have specific limited values. For example makeExampleSeSAMeDataSet , it seems like the value of platform can only be HM450, EPIC or HM27 so having the arugment platform = c("HM450", "EPIC", "HM27") and then use platform <- match.arg(platform)` in the code . More general, provide some initial checking if the input to your main functions is the expected object type. This should be applied throughout the entire code. (ie. stopifnot)

_Thanks for the suggestion. I went through the entire code base and adopted match.arg in 21 places (mostly for platform and refversion). 24 new stopifnot were added to perform initial check of the function input._

**Comment by lshep** [comment]
I'm going to manually rerun both packages for builds -  it should allow the software package to find the data package -

**Comment by lshep** [comment]
The sesame package looks in good shape and I have no further comments regarding the software package. 

The sesameData package is still somewhat concerning regarding the size of the package.  It contains a lot of data making it very substantial (and possibly time consuming) for a user to download.   We highly recommend reformatting to make this an experiment package that utilizes experimenthub.  The benefit being the package becomes a light weight as the data will be stored in Amazon S3 buckets and then you and the users can download data as necessary for use.  There is a guideline for experimenthub packages found [here](https://github.com/Bioconductor/ExperimentHubData/blob/master/vignettes/CreateAnExperimentHubPackage.Rmd).  Note the R/ section is optional. 

It is very similar to what you already have the differences being
1.  The data will be removed from the package and uploaded to our S3 bucket (I will send you credentials to the email you have utilized to reach out to me) - you will have to use `git rm` to remove the data and possibly clean the repository which I can send instructions if necessary - 
2.  Create a inst/extdata/metadata.csv  with required fields for each item you upload to S3.  To see if all required columns are present you can use `ExperimentHubData::makeExperimentHubMetadata("\<pathto>\sesameData")` in an R session that will test reading the metadata.csv. 
3.  Create R functions that wrap around the downloading of data from experimenthub
4.  Update to depend on experimenthub 
5.  Update vignette as necessary

I can assist with any questions you have regarding this as I am Bioconductor team member that currently manages the hubs. 

It looks like there was data pertaining to certain experiments (EPIC, HM27, HM450, etc) - You can either upload each individual file to the hub or if generally all these files are needed for analysis you could save those objects pertaining to the particular type in one .rda file to be uploaded to minimize effort - I don't know the specifics of how these files are utilized so this may not be how they are intended and all individual may be necessary - 
 

Cheers, 
Lori

**Comment by zwdzwd** [comment]
Thank you for the tips @lshep . The reason I defined a new environment is to avoid repeated loading from disk. For example, if a function that loads the data is called again (in fact some function may be repeated many many times), I want it to reuse the data the same function loaded. Is there a better way to do that with just ExperimentHub? That would be great.

**Comment by lshep** [comment]
I'm looking at your packages now and will have further feedback but to answer: 

"The accessor functions seem to be loading the data from the disk again everytime it's called. Is there a way to keep an internal copy in memory (just as the environment in my implementation) and return the copy every time the accessor function is called?"

The ExperimentHub returns the object so you could assign in a variable or global variable name -  is this what you mean?

**Comment by lshep** [comment]
The version bump required was from me kicking off the builds manually once the SPB switched to using Bioc 3.8 so I removed the tag.  

I will have a review for you shortly.

**Comment by lshep** [comment]
**SesameData Package:**

**BuildReport:**

- [ ] [Required]  documentation WARNING - Becaues you added the .onLoad individual functions have
been made for each of the data.  So you could do something like
```
library(sesameData)
EPIC.1.LNCaP = EPIC.1.LNCaP()
```
So this is saying those functions need to be documented.  

- [ ]  [Optional but strongly encouraged] I would still add unit tests - it could be  testing that the  sesameDataList returns the expected number of files -  testing one or two downloads that the expected class is returned - ect. 

- [ ] [Optional but strongly encouraged] Add a NEWS file to the top level package directory. This is where you could announce changes to functions or data and are announced at release time. 

**R code:** 
zzz.R   
- [ ] [Required] Please update `message()` function to `packageStartupMessage()` function
- [ ] [Required] Please remove the 
```
suppressMessages(log <- capture.output(
        sesameDataCacheAll()));
```
Essentially you are making each user of the data package download all of your data. The data may eventually be used in other ways you didn't intend and the user shouldn't be forced to download all of the files unless they want or unless necessary. 

data.R
- [ ] [Quetstion/Discussion] I still don't think you would need environment - if you have a separate function for loading vs processing and assign the data to a variable this would be better than reloading or the potential to reload multiple times anyways - And regardless I don't think its doing what you want - 

- [ ] [Quetstion/Discussion] Do you need the sesameDataGet functions now that you have the individual
functions with the .onLoad? 

- [ ] [Comment]  I like the sesameDataList function - I haven't seen that before and its nifty for knowing
what the user has access to especially since you added the .onLoad.

**inst/extdata:**
Where are these files being utilized?  I can find you actually downloading or analyzing the GSM3040017* files in either the data package or the software package.  These files should be removed which will also greatly decrease the size of your package. Please do `git rm` on the GSM3040017* files and http://bioconductor.org/developers/how-to/git/remove-large-data/ and commit changes. 

I see that the 4207113116* files are utilized in sesame so these are fine. 

**Vignette:** 
- [ ] [Required] We still would require a minimal vignette.  I would recommended describing the data that is available and that it is designed to be utilized with the sesame package.  Maybe minimal code example for downloading data like 
```
# The available data can be shown with:
sesameDataList()

# Each has its own download function for example:
EPIC.1.LNCaP = EPIC.1.LNCaP()
```

**Description:**
- [ ] Remove `LazyData: TRUE`. This rarely proves to be a good thing. In our experience it only slows down the loading of packages with large data.

&nbsp;&nbsp;
&nbsp;&nbsp;
&nbsp;&nbsp;
&nbsp;&nbsp;

**Sesame Software:**
Thank you for implementing the other previously suggested improvements from val.  Some minor additional suggestions: 

**README:**
- [ ] [Optional but strongly encouraged] Add a section for installation from Bioconductor - can still include the install_github as a "development" version"
```
source("https://bioconductor.org/biocLite.R")
biocLite("sesame")
```

**DESCRIPTION:**
- [ ] [Optional but strongly encouraged] Remove `LazyData: TRUE`. This rarely proves to be a good thing. In our experience it only slows down the loading of packages with large data.

**R code:**
- [ ] As you have designed the S4 class,  it would be better to implement getter/setter methods for the object internals and to use those throughtout your code rather than the `@` 
ex. (these would need to be expanded, possibly check for right formatting, maybe
additional arguments depending on struture, but you get the basic idea )
```
setMethod("IG", "sigSet", function(x){ x@IG })
setReplacementMethod("IG", "sigSet", function(x, value){ x@IG <- value})

```
&nbsp;&nbsp;
&nbsp;&nbsp;
&nbsp;&nbsp;
&nbsp;&nbsp;

I think the packages are in a very good state and with these minor improvements be close to ready for acceptance.  
Please implement changes and when ready for a re-review initialize a version bump in the DESCRIPTION for new build reports and comment back here with updates. 

Cheers

**Comment by lshep** [comment]
You might also find this thread interesting:  https://support.bioconductor.org/p/108995/#108998

**Comment by lshep** [comment]
Hi @zwdzwd  Just wanted to check it to see how this was coming along?

**Comment by zwdzwd** [comment]
Thank you for getting the reviews back Lori. Sorry that I had been swamped by a few other things in the past week (a paper deadline and a travel) and is just getting back to this. I read your review and I concur with most of the things you mentioned. But I’d like to clarify on three comments you raised and hear your opinion before I move forward to commit changes.

> 1. [Quetstion/Discussion] I still don't think you would need environment - if you have a separate function for loading vs processing and assign the data to a variable this would be better than reloading or the potential to reload multiple times anyways - And regardless I don't think its doing what you want -

I now think there is perhaps a fundamental difference between how ExperimentHub was designed and how I intended to use it. As the name suggests, ExperimentHub was intended to be used as a source to load **test data** or **experiment data** which is supposed to be used by the users directly. For example, a user could load an “experiment” and start to work on it by applying functions that operates on the experiment. But in my case, I intend to use the data as **internal data**. I do not want to expose the usage of these data to the user. For example, a data can be about the internal design information about a platform, some package function should handle this data loading instead of having the users  manually load every internal data and feed to the function (which will always use the data). 

I have indeed considered having separate functions for data loading and processing. But it’s surely damaging the function abstraction idea if the loaded data is for internal use. And in fact, it can be tedious too if there are multiple data that needs be used and the user need to load the independently and feed it to the function that process it.

In other words, I guess what I am getting at can be abstracted into the following equally valid implementation:

implementation1: 
`foo1(input_data, internal_data) {actual_work}`
implementation2: 
`foo2(input_data) {load_internal_data; actual_work}`

I prefer to use foo2 implementation for the internal_data since it does the function abstraction better.

>     [Required] Please remove the suppressMessages(log <- capture.output(     
>     sesameDataCacheAll()));


Related to the same complication mentioned above, there is an important side issue related to parallel processing: calling foo2 in parallel means calling load_internal_data in parallel. This will lead to malfunctioning when the data was not cached and needs to be fetched from internet, as I previously mentioned the in the [ExperimentHub issue](https://github.com/Bioconductor/ExperimentHub/issues/2). So I decided to have all the data cached to disk at the time when the package is loaded. I know it's not an ideal solution since this violates the purpose of ExperimentHub that should allow data be cached or loaded only upon needs. But it's at least better than seeing the package not functioning in parallel. This explains why I implemented sesameCacheAll() under zzz.R. 

> You might also find this thread interesting: https://support.bioconductor.org/p/108995/#108998

Interestingly, I actually I find the top response in this thread you forwarded doing exactly the same thing as I was doing in sesameData package, for the same purpose—reducing disk access from ExperimentHub call. The only difference is in the response of the thread, the parent environment is emptyenv() and in my implementation the environment is under the sesameData package namespace. 

I hope you see my struggle. I made multiple attempts to incorporate ExperimentHub, balancing my development goal and the design purpose of ExperimentHub. And this may have already been shared by other users who tried to use ExperimentHub for internal data as well. Please let me know your insights/recommendation or if I missed some important solutions.

Thank you,

**Comment by lshep** [comment]
Yes that makes sense.  We just want to make sure there is sound reasoning behind decisions.  Please proceed with the other minor changes but this may be left as implemented.  Cheers.

**Comment by zwdzwd** [comment]
Thanks so much for the suggestion on these good practices @lshep . I have altered my implementation following your guide:

- Functions attached onLoad
I ended up removing these attachment and stick with `sesameDataGet` as mentioned, this way we also avoid some namespace pollution. I did add unit tests as suggested for every loadable data.

- NEWS file
Added.

- Please do git rm on the GSM3040017* files and http://bioconductor.org/developers/how-to/git/remove-large-data/ and commit changes.
I removed GSM3040017*. This was actually used implicitly with directory listing (so the file name didn't appear in the example). But for clarity, I removed the files as suggested. 

- Vignette:
I added a very simple vignette as suggested.

-  Remove LazyData: TRUE. 
Done.

- README:  [Optional but strongly encouraged] Add a section for installation from Bioconductor - can still include the install_github as a "development" version"
Done

- [Optional but strongly encouraged] Remove LazyData: TRUE
Done

- getter and setter methods
These are now implemented and used throughout the code, as suggested.

Thank you @lshep , let me know if you see other improvements.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/zwdzwd.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"sesameData\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/sesameData

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/zwdzwd.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"sesame\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/sesame

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
